README
This is Perl module Amon2::Setup::Flavor::Teng.

INSTALLATION

Amon2::Setup::Flavor::Teng installation is straightforward. If your CPAN shell is set up,
you should just be able to do

% cpan Amon2::Setup::Flavor::Teng

Download it, unpack it, then build it as per the usual:

% perl Makefile.PL
% make && make test

Then install it:

% make install

DOCUMENTATION

Amon2::Setup::Flavor::Teng documentation is available as in POD. So you can do:

% perldoc Amon2::Setup::Flavor::Teng


Changes
Revision history for Perl extension Amon2::Setup::Flavor::Teng

{{$NEXT}}

0.05 2013-04-20T14:19:20Z
	    [Improvements]
        - Minillaize
        [Bug Fixes]
        - fix typo (yoshidaster)

0.04    Tue Aug 21 19:23:21 2012
       	[Improvements]
        - correspond to Teng 0.05 (memememomo)
        [Bug Fixes]
        - add dependency Amon2::Plugin::DBI (tokuhirom)

0.03    Mon Nov 21 10:18:32 2011
        [Bug Fixes]
        - fix 'Query was empty' error in mysql. (tokuhirom)
        [Improvements]
        - cache instance of schema object.      (tokuhirom)
        - create lib/<<PATH>>/DB.pm by default for extensibility. (tokuhirom)

0.02    Mon Oct 28 12:41:23 2011
        [Bug Fixes]
        - fixed generated tests.
          tests failed if sessions table exists.

0.01    Mon Oct 24 16:59:03 2011
        - original version


lib/Amon2/Setup/Flavor/TengWithoutLoader.pm
use strict;
use warnings;
use utf8;

package Amon2::Setup::Flavor::TengWithoutLoader;
use parent qw(Amon2::Setup::Flavor);

sub run {
    my $self = shift;

    $self->write_file('lib/<<PATH>>.pm', <<'...');
package <% $module %>;
use strict;
use warnings;
use utf8;
use parent qw/Amon2/;
use 5.008001;

__PACKAGE__->load_plugin(qw/DBI/);

# initialize database
use DBI;
sub setup_schema {
    my $self = shift;
    my $dbh = $self->dbh();
    my $driver_name = $dbh->{Driver}->{Name};
    my $fname = lc("sql/${driver_name}.sql");
    open my $fh, '<:encoding(UTF-8)', $fname or die "$fname: $!";
    my $source = do { local $/; <$fh> };
	for my $stmt (split /;/, $source) {
		$dbh->do($stmt) or die $dbh->errstr();
	}
}

use <% $module %>::DB;
sub db {
    my $self = shift;
    if ( !defined $self->{db} ) {
        my $conf = $self->config->{'DBI'}
        or die "missing configuration for 'DBI'";
        my $dbh = DBI->connect(@{$conf});
        $self->{db} = <% $module %>::DB->new(
            dbh    => $dbh,
	    );
    }
    return $self->{db};
}

1;
...

    $self->write_file('lib/<<PATH>>/DB.pm', <<'...');
package <% $module %>::DB;
use parent qw(Teng);
1;
...

    $self->write_file('lib/<<PATH>>/DB/Schema.pm', <<'...');
package <% $module %>::DB::Schema;
use Teng::Schema::Declare;

table {
    name 'sessions';
    pk 'id';
    columns qw(id session_data);
};

1;
...

    $self->write_file('t/09_teng_without_loader.t', <<'...');
use strict;
use warnings;
use Test::More;
use <% $module %>;

my $teng = <% $module %>->new;
is(ref $teng, '<% $module %>', 'instance');
is(ref $teng->db, 'My::App::DB', 'instance');
$teng->db->do("create table if not exists sessions (id char(72) primary key, session_data text)") or die $teng->db->errstr;
$teng->db->insert('sessions', { id => 'abcdefghijklmnopqrstuvwxyz', session_data => 'ka2u' });
my $res = $teng->db->single('sessions', { id => 'abcdefghijklmnopqrstuvwxyz' });
is($res->get_column('session_data'), 'ka2u', 'search');
$teng->db->delete('sessions', {id => 'abcdefghijklmnopqrstuvwxyz'});

done_testing;
...
}

1;

__END__

=encoding utf-8

=head1 NAME

Amon2::Setup::Flavor::TengWithoutLoader - Teng Schema Flavor for Amon2

=head1 SYNOPSIS

    amon2-setup.pl --flavor Basic,TengWithoutLoader My::App

=head1 DESCRIPTION

Easy setup Teng ORM for Amon2.
This doesn't use Teng::Schema::Loader, create schema class.

=head1 AUTHOR

Kazuhiro Shibuya

=head1 SEE ALSO

Amon2 L<http://search.cpan.org/~tokuhirom/Amon2/> , Teng L<http://search.cpan.org/~nekokak/Teng/>

=head1 LICENSE

Copyright (C) Kazuhiro Shibuya

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Amon2/Setup/Flavor/Teng.pm
use strict;
use warnings;
use utf8;

package Amon2::Setup::Flavor::Teng;
use parent qw(Amon2::Setup::Flavor);
our $VERSION = '0.05';

sub run {
    my $self = shift;

    $self->write_file('lib/<<PATH>>.pm', <<'...');
package <% $module %>;
use strict;
use warnings;
use utf8;
use parent qw/Amon2/;
our $VERSION='0.01';
use 5.008001;

__PACKAGE__->load_plugin(qw/DBI/);

# initialize database
use DBI;
sub setup_schema {
    my $self = shift;
    my $dbh = $self->dbh();
    my $driver_name = $dbh->{Driver}->{Name};
    my $fname = lc("sql/${driver_name}.sql");
    open my $fh, '<:encoding(UTF-8)', $fname or die "$fname: $!";
    my $source = do { local $/; <$fh> };
    for my $stmt (split /;/, $source) {
        next unless $stmt =~ /\S/;
        $dbh->do($stmt) or die $dbh->errstr();
    }
}

use Teng;
use Teng::Schema::Loader;
use <% $module %>::DB;
my $schema;
sub db {
    my $self = shift;
    if ( !defined $self->{db} ) {
        my $conf = $self->config->{'DBI'}
        or die "missing configuration for 'DBI'";
        my $dbh = DBI->connect(@{$conf});
        if ( !defined $schema ) {
            $self->{db} = Teng::Schema::Loader->load(
                namespace => '<% $module %>::DB',
                dbh       => $dbh,
	    );
            $schema = $self->{db}->schema;
        } else {
            $self->{db} = <% $module %>::DB->new(
                dbh    => $dbh,
                schema => $schema,
	    );
        }
    }
    return $self->{db};
}

1;
...

    $self->write_file('lib/<<PATH>>/DB.pm', <<'...');
package <% $module %>::DB;
use strict;
use warnings;
use utf8;
use parent qw(Teng);

1;
...

    $self->write_file('t/08_teng.t', <<'...');
use strict;
use warnings;
use DBI;
use Test::More;
use <% $module %>;

my $dbi = DBI->connect('dbi:SQLite:dbname=db/development.db');
$dbi->do("create table if not exists sessions (id char(72) primary key, session_data text)") or die $dbi->errstr;
my $teng = <% $module %>->new;
is(ref $teng, '<% $module %>', 'instance');
isa_ok($teng->db, 'Teng', 'instance');
isa_ok($teng->db, '<% $module %>::DB', 'instance');
$teng->db->insert('sessions', { id => 'abcdefghijklmnopqrstuvwxyz', session_data => 'ka2u' });
my $res = $teng->db->single('sessions', { id => 'abcdefghijklmnopqrstuvwxyz' });
is($res->get_column('session_data'), 'ka2u', 'search');
$teng->db->delete('sessions', {id => 'abcdefghijklmnopqrstuvwxyz'});

done_testing;
...
}

1;

__END__

=encoding utf-8

=head1 NAME

Amon2::Setup::Flavor::Teng - Teng flavor for Amon2

=head1 SYNOPSIS

    amon2-setup.pl --flavor Basic,Teng My::App

=head1 DESCRIPTION

Easy setup Teng ORM for Amon2.

=head1 AUTHOR

Kazuhiro Shibuya

=head1 SEE ALSO

Amon2 L<http://search.cpan.org/~tokuhirom/Amon2/> , Teng L<http://search.cpan.org/~nekokak/Teng/>

=head1 LICENSE

Copyright (C) Kazuhiro Shibuya

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
